### PR TITLE
Create separate functions for df as edge-lists and adjacency matrices

### LINKED ISSUE #2464: to_pandas_dataframe is not the inverse of from_pandas_dataframe
The function `from_pandas_dataframe` assumes that the data frame has edge-list like structures, but `to_pandas_dataframe` generates an adjacency matrix. 

Personally, I think the edge-list like structures are more useful because:
(1) Otherwise these two functions are just (wrapped) duplications of `numpy_matrix` related functions
(2) The edge-list like data frames are convenient for other interfaces such as `py2cytoscape`. 

Some related commits:
[52eb3d3](https://github.com/networkx/networkx/commit/52eb3d38ff78e4002a424ae5e8976998afdaf1a2): `from_pandas_dataframe` and `to_pandas_dataframe` were added. At that time both functions were about adjacency matrix.  
[f0c22dd](https://github.com/networkx/networkx/commit/f0c22dd3305ff236860c651d8a08fbabcb8b0a77): `from_pandas_dataframe` was changed to current behavior, but it seems that `to_pandas_dataframe` was forgotten to be changed accordingly.

### PR DESCRIPTION
**Pull request by jarrodmillman**
Fixes #2464.

Some related commits:
52eb3d3: from_pandas_dataframe and to_pandas_dataframe were added to work with adjacency matrix.
f0c22dd: from_pandas_dataframe was changed to work with edge-lists, but to_pandas_dataframe was not changed.

### CODE REVIEW

File: networkx/tests/test_convert_pandas.py [hunk 1]
```diff
@@ -34,40 +34,69 @@ def __init__(self, ):
     def assert_equal(self, G1, G2):
```
Comment (jarrodmillman) [inline networkx/tests/test_convert_pandas.py]: Should we use ``assert_graphs_equal`` instead?

Comment (dschult) [inline networkx/tests/test_convert_pandas.py]: ```is_isomorphic``` is much harder to test than what our ```assert_graphs_equal``` tests.
It looks like the current code doesn't actually need the is_isomorphic though.  
Those graphs really are the same -- not just isomorphic.  If that's true, then yes -- please use ```assert_graphs_equal```.

File: networkx/convert_matrix.py [hunk 2]
```diff
@@ -129,8 +143,123 @@ def to_pandas_dataframe(G, nodelist=None, dtype=None, order=None,
...
+
+    >>> import numpy as np
+    >>> import pandas as pd
+    >>> df = pd.DataFrame([[1, 1], [2, 1]])
+    >>> df
+       0  1
+    0  1  1
+    1  2  1
+    >>> G = nx.from_pandas_adjacency(df)
+    >>> nx.info(G)
```
Comment (jarrodmillman) [inline networkx/convert_matrix.py]: I tried using ``print(nx.info(G))``to make the output nicer.  But I was getting an error presumably due to spacing issues.  Thoughts?

File: networkx/convert_matrix.py [hunk 3]
```diff
@@ -129,8 +143,123 @@ def to_pandas_dataframe(G, nodelist=None, dtype=None, order=None,
...
+
+    """
+    edgelist = nx.to_edgelist(G)
+    source = [s for s, t, d in edgelist]
+    target = [t for s, t, d in edgelist]
+    all_keys = set().union(*(d.keys() for s, t, d in edgelist))
+    edge_attr = {k: [d.get(k, float("nan")) for s, t, d in edgelist] for k in all_keys}
+    edgelistdict = {'source': source, 'target': target}
+    edgelistdict.update(edge_attr)
+    return pd.DataFrame(edgelistdict)
```
Comment (jarrodmillman) [inline networkx/convert_matrix.py]: This is just a sketch.  I wanted to get some feedback before I put too much effort into this.  Does this seem like a reasonable approach?  If so, I will implement this more fully for more substantial feedback.

File: networkx/convert_matrix.py [hunk 4]
```diff
@@ -20,29 +26,37 @@
...
+    """
+    msg = "to_pandas_dataframe is deprecated and will be removed" \
+        "in 2.1, use to_pandas_adjacency instead."
+    _warnings.warn(msg, DeprecationWarning)
+
+    return to_pandas_adjacency(G, nodelist=None, dtype=None, order=None,
+                               multigraph_weight=sum, weight='weight', nonedge=0.0)
+
+
+def to_pandas_adjacency(G, nodelist=None, dtype=None, order=None,
```
Comment (jarrodmillman) [inline networkx/convert_matrix.py]: TODO: document ``dtype`` and ``order``.

File: networkx/convert_matrix.py (hunk 3 above)
Comment (dschult) [inline networkx/convert_matrix.py]: Yup -- I think this is what is needed.
You should just use ```G.edges(data=True)``` instead of ```nx.to_edgelist(G)```

**Comment by jarrodmillman** [comment]
@dschult Would you review this?  In particular, here is how I checked that both ``to``/``from`` pairs would provide a roundtrip:
```
    def test_roundtrip(self):

        # edgelist
        Gtrue = nx.Graph([(1, 1), (1, 2)])
        df = nx.to_pandas_edgelist(Gtrue)
        G = nx.from_pandas_edgelist(df)
        assert_graphs_equal(Gtrue, G)

        # adjacency
        Gtrue = nx.Graph(({1: {1: {'weight': 1}, 2: {'weight': 1}}, 2: {1: {'weight': 1}}}))
        df = nx.to_pandas_adjacency(Gtrue, dtype=int)
        G = nx.from_pandas_adjacency(df)
        assert_graphs_equal(Gtrue, G)
```

Note that if I did
```
>>> Gtrue = nx.Graph([(1, 1), (1, 2)])
>>> df = nx.to_pandas_adjacency(Gtrue, dtype=int)
>>> G = nx.from_pandas_adjacency(df)
```
then
```
>>> Gtrue.edges(data=True)
EdgeDataView([(1, 1, {}), (1, 2, {})])
>>> G.edges(data=True)
EdgeDataView([(1, 1, {'weight': 1}), (1, 2, {'weight': 1})])
```

I am not sure that is what we want.

Other than that I don't have an other plans for the code.  You can ignore the fact that Travis is failing (for now) because it is due to conda not supporting Python 3.4 recently.  So the newest version of pandas they have is quite old.  I am working on a PR related to this, which I think will solve the problem.  So I will wait for that to be sorted out before I finalize this.  But I don't anticipate changing the API.  So let me know if this seems ready to merge once I sort out the old pandas in conda on python 3.4 issue.

**Comment by dschult** [comment]
Hmmm...  Why does ```to_pandas_edgelist``` add the edge attribute ```weight```?
I would expect that 

    >>> Gtrue = nx.Graph([(1, 1), (1, 2)])
    >>> df = nx.to_pandas_edgelist(Gtrue, dtype=int)
    
would create a pandas frame without any ```weight``` column.  Just ```source``` and ```target```.  I understand that this is not the API that was in place before this PR, but I think it would be better.  Ideally you could specify a list of edge attributes that would each create a column in the dataframe. If the list is empty (or maybe None?) then no edge attribute columns are written. This is how we read the dataframes so we should write them that way too.  I would be fine with that as a separate PR though.

I think the way you are testing the round trip is good.

**Pull request merged**